dask_match/io/parquet.py

```diff
@@ -0,0 +1,330 @@
+    get_engine,
+    process_statistics,
+    set_index_columns,
+    ParquetFunctionWrapper,
+)
+from dask.dataframe.io.parquet.utils import _split_user_options
+from dask.utils import natural_sort_key
+from matchpy import (
+    CustomConstraint,
+    Pattern,
+    ReplacementRule,
+    Wildcard,
+)
+
+from dask_match.core import IO, LE, LT, GE, GT, EQ, NE, Filter
+
+
+NONE_LABEL = "__null_dask_index__"
+
+
+class ReadParquet(IO):
+    """Read a parquet dataset"""
+
+    _parameters = [
+        "path",
+        "columns",
+        "filters",
+        "categories",
+        "index",
+        "storage_options",
+        "engine",
+        "use_nullable_dtypes",
+        "calculate_divisions",
+        "ignore_metadata_file",
+        "metadata_task_size",
+        "split_row_groups",
+        "blocksize",
+        "aggregate_files",
+        "parquet_file_extension",
+        "filesystem",
+        "kwargs",
+    ]
+    _defaults = {
+        "columns": None,
+        "filters": None,
+        "categories": None,
+        "index": None,
+        "storage_options": None,
+        "engine": "auto",
+        "use_nullable_dtypes": False,
+        "calculate_divisions": False,
+        "ignore_metadata_file": False,
+        "metadata_task_size": None,
+        "split_row_groups": "infer",
+        "blocksize": "default",
+        "aggregate_files": None,
+        "parquet_file_extension": (".parq", ".parquet", ".pq"),
+        "filesystem": None,
+        "kwargs": {},
+    }
+
+    @staticmethod
+    def normalize(
+        path=None,
+        columns=None,
+        filters=None,
+        categories=None,
+        index=None,
+        storage_options=None,
+        engine="auto",
+        use_nullable_dtypes=False,
+        calculate_divisions=False,
+        ignore_metadata_file=False,
+        metadata_task_size=None,
+        split_row_groups="infer",
+        blocksize="default",
+        aggregate_files=None,
+        parquet_file_extension=(".parq", ".parquet", ".pq"),
+        filesystem=None,
+        **kwargs,
+    ):
+        if isinstance(columns, (str, int)):
+            columns = [columns]
+        elif isinstance(columns, tuple):
+            columns = list(columns)
+
+        if use_nullable_dtypes:
+            use_nullable_dtypes = dask.config.get("dataframe.dtype_backend")
+
+        if isinstance(engine, str):
+            engine = get_engine(engine)
+
+        if hasattr(path, "name"):
+            path = stringify_path(path)
+
+        return (
+            path,
+            columns,
+            filters,
+            categories,
+            index,
+            storage_options,
+            engine,
+            use_nullable_dtypes,
+            calculate_divisions,
+            ignore_metadata_file,
+            metadata_task_size,
+            split_row_groups,
+            blocksize,
+            aggregate_files,
+            parquet_file_extension,
+            filesystem,
+            kwargs,
+        ), {}
```

Comment:
I'm thinking of removing this method and instead relying on the `def read_parquet` and similar functions.  The classes won't do any normalization, and we'll depend on the user-facing API to handle it.  Thoughts?

Reply:
I think that would be fine. We definitely need a space to "normalize" user inputs, but don't need to do this within the `API` subclasses themselves.

Reply:
coming a bit late to the party but strong +1 for doing normalization/validation of inputs far up in the stack

Reply:
This is done in main now.  I've removed the normalize function.  So far I'm still comfortable with Exprs not having `__init__` methods.